### PR TITLE
[Bugfix] Fix multigpu `dispatch_for_generation`

### PR DESCRIPTION
**Pull request by kylesayrs**
## Purpose ##
* Fix `test_oneshot_and_finetune_with_tokenizer.py` with multiple cuda devices
  * This test had two failures, the first failure occurred as a result of sequential onloading introducing `dispatch_for_generation`. This function is also used to dispatch for training. However, this method did not account for no split modules
  * The second failure is an existing failure where HFTrainer.compute_loss does not account for multi-gpu models. This will be fixed in the next release by https://github.com/huggingface/transformers/pull/38029

## Changes ##
* Pass no split modules when computing a device map for generation (and training)
* Load model on CPU (since this is now the default flow as of sequential onloading landing

## Testing ##
* Ran `test_oneshot_and_finetune_with_tokenizer` with two GPUs to completion (with upstream transformers)

### CODE REVIEW

File: src/llmcompressor/transformers/finetune/session_mixin.py [hunk 1]
```diff
@@ -286,6 +286,8 @@ def compute_loss(
 
         # TODO: do we need these model signature columns?
         inputs = {k: inputs[k] for k in inputs if k in self._signature_columns}
+        if num_items_in_batch is not None:
+            num_items_in_batch = num_items_in_batch.item()
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/transformers/finetune/session_mixin.py]: Yep

File: src/llmcompressor/utils/dev.py [hunk 2]
```diff
@@ -126,11 +126,17 @@ def dispatch_for_generation(model: PreTrainedModel) -> PreTrainedModel:
...
         model,
         dtype=model.dtype,
-        no_split_module_classes=model._get_no_split_modules("auto"),
+        no_split_module_classes=no_split_module_classes,
+    )
+    device_map = infer_auto_device_map(
+        model,
+        dtype=model.dtype,
+        max_memory=max_memory,
+        no_split_module_classes=no_split_module_classes,
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/utils/dev.py:139]: Yep

File: tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def test_oneshot_and_finetune_with_tokenizer(self):
             self.model,
         )
         model_loaded = AutoModelForCausalLM.from_pretrained(
-            self.model, device_map="cuda:0", torch_dtype="auto"
+            self.model, torch_dtype="auto"
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py:39]: Yep

File: src/llmcompressor/transformers/finetune/session_mixin.py [hunk 4]
```diff
@@ -286,6 +286,8 @@ def compute_loss(
...
         loss = super().compute_loss(
```
Comment (brian-dellabetta) [inline src/llmcompressor/transformers/finetune/session_mixin.py]: this is my first time looking into `SessionManagerMixIn`, how does this work? what is `super()` when the class doesn't extend anything? The docstring says it extends the Hugging Face Trainer class, but it doesn't do so in the code

**Review by dsikka — APPROVED** [review]
Can confirm that this fixes the other two nightly errors as well:

`tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune.py`
and  `tests/llmcompressor/transformers/finetune/test_finetune_no_recipe_custom_dataset.py`

File: src/llmcompressor/transformers/finetune/session_mixin.py (hunk 4 above)
Comment (kylesayrs) [reply src/llmcompressor/transformers/finetune/session_mixin.py]: This design was meant to integrate "arbitrary" trainers, ie either the HFTrainer or the SFTTrainer. See `src/llmcompressor/transformers/finetune/trainer.py`

**Review by rahul-tuli — COMMENTED** [review]
Blazing fast resolution!

**Comment by kylesayrs** [comment]
@dsikka I've confirmed that these tests run until the original `loss = loss / num_items_in_batch` issue, which is fixed in upstream transformers
